xapian-core/cluster/kmeans.cc

```diff
@@ -61,22 +70,85 @@ KMeans::initialize_centroids(ClusterSet &cset) {
 }
 
 void
-KMeans::initialize_points(MSetDocumentSource source, TermListGroup tlg) {
-    while(!source.at_end()) {
+KMeans::initialize_kmeanspp(ClusterSet &cset) {
```

Comment:
Another pre-clustering option we might want to explore (in the future, not within GSoC) is [canopy clustering](https://en.wikipedia.org/wiki/Canopy_clustering_algorithm), and in deciding on the API for using KMeans++ it may be worth thinking about what happens when you have KMeans++, PSO and Canopy as different pre-clusterers.